packages/react-dom/src/events/ReactDOMEventListener.js

```diff
@@ -125,70 +137,108 @@ export function isEnabled() {
     ? dispatchInteractiveEvent
     : dispatchEvent;
+  const rawEventName = getRawEventName(topLevelType);
 
-  addEventBubbleListener(
+  trapEvent(
+    addEventBubbleListener,
     element,
-    getRawEventName(topLevelType),
-    // Check if interactive and wrap in interactiveUpdates
-    dispatch.bind(null, topLevelType),
+    topLevelType,
+    dispatch,
+    rawEventName,
+    isLegacy,
   );
 }
 
-/**
- * Traps a top-level event by using event capturing.
- *
- * @param {number} topLevelType Number from `TopLevelEventTypes`.
- * @param {object} element Element on which to attach listener.
- * @return {?object} An object with a remove function which will forcefully
- *                  remove the listener.
- * @internal
- */
 export function trapCapturedEvent(
   topLevelType: DOMTopLevelEventType,
-  element: Document | Element,
-) {
-  if (!element) {
-    return null;
-  }
+  element: Document | Element | Node,
+  isLegacy: boolean,
+): void {
   const dispatch = isInteractiveTopLevelEventType(topLevelType)
     ? dispatchInteractiveEvent
     : dispatchEvent;
+  const rawEventName = getRawEventName(topLevelType);
 
-  addEventCaptureListener(
+  trapEvent(
+    addEventCaptureListener,
     element,
-    getRawEventName(topLevelType),
-    // Check if interactive and wrap in interactiveUpdates
-    dispatch.bind(null, topLevelType),
+    topLevelType,
+    dispatch,
+    rawEventName,
+    isLegacy,
   );
 }
 
-function dispatchInteractiveEvent(topLevelType, nativeEvent) {
-  interactiveUpdates(dispatchEvent, topLevelType, nativeEvent);
+type Dispatcher = (
+  topLevelType: DOMTopLevelEventType,
+  passive: null | boolean,
+  nativeEvent: AnyNativeEvent,
+) => void;
+
+// A helper function to remove bytes
+function bindDispatch(
+  dispatch: Dispatcher,
+  topLevelType: DOMTopLevelEventType,
+  passive: null | boolean,
+) {
+  return dispatch.bind(null, topLevelType, passive);
+}
+
+function trapEvent(
+  eventListener: (
+    element: Document | Element | Node,
+    eventName: string,
+    listener: (event: AnyNativeEvent) => void,
+    passive: boolean | null,
+  ) => void,
+  element: Document | Element | Node,
+  topLevelType: DOMTopLevelEventType,
+  dispatch: Dispatcher,
+  rawEventName: string,
+  isLegacy: boolean,
+) {
+  if (isLegacy) {
+    // Check if interactive and wrap in interactiveUpdates
+    const listener = bindDispatch(dispatch, topLevelType, null);
+    // We don't listen for passive/non-passive
+    eventListener(element, rawEventName, listener, null);
+  } else {
+    if (passiveBrowserEventsSupported) {
+      // Check if interactive and wrap in interactiveUpdates
+      const activeListener = bindDispatch(dispatch, topLevelType, false);
+      const passiveListener = bindDispatch(dispatch, topLevelType, true);
+      // We listen to the same event for both passive/non-passive
+      eventListener(element, rawEventName, passiveListener, true);
+      eventListener(element, rawEventName, activeListener, false);
```

Comment:
Can you explain more about why we listen for both? Sorry I'm not quite following.

Reply:
We listen to both so we can handle both cases in future plugins event APIs. This is because we listen on the root ahead of time so we need both to handle specific logic that is only possible in one or the other